### PR TITLE
feat(parser)!: Support ALTER VIEW AS SELECT

### PR DESCRIPTION
**Pull request by xiaohui-sun**
Add a support to "ALTER VIEW AS SELECT"

`ALTER VIEW view_identifier AS select_statement`

More documentation on it:
https://spark.apache.org/docs/3.0.0-preview2/sql-ref-syntax-ddl-alter-view.html#syntax-3

### CODE REVIEW

File: sqlglot/expressions.py
Comment (georgesittas) [inline sqlglot/expressions.py:4289]: Not sure if this is how we wanna implement this. For `CREATE ...` statements we have a single `Create` node which contains a `kind` argument that stores `TABLE`, `VIEW` etc. Maybe we want to do the same for `Alter` statements.

File: sqlglot/parser.py
Comment (georgesittas) [inline sqlglot/parser.py:6488]: Maybe we can instead make a set of tokens, `ALTERABLES`, starting with `TABLE` and `VIEW` and then do:

```python
if not self._match_set(self.ALTERABLES):
    return self._parse_as_command(start)
...
```

This way we'll avoid adding the `start` argument everywhere.

File: sqlglot/parser.py
Comment (georgesittas) [inline sqlglot/parser.py:6516]: This needs to be cleaned up

File: sqlglot/parser.py
Comment (georgesittas) [inline sqlglot/parser.py:6518]: I think we can consolidate these two, imitating the current implementation for `ALTER TABLE`:

```suggestion
        this = self._parse_table(schema=True)
```

File: sqlglot/parser.py
Comment (georgesittas) [inline sqlglot/parser.py:6537]:
```suggestion
        if self._match(TokenType.VIEW):
            return self._parse_alter_view(start)
        return self._parse_as_command(start)
```

**Review by georgesittas — COMMENTED** [review]
Thanks for the PR @xiaohui-sun 👍

Please run `make style` to fix the styling CI errors.

File: sqlglot/expressions.py
Comment (xiaohui-sun) [reply sqlglot/expressions.py:4289]: Thanks for reviewing!
Yes that makes sense. Let me add that.

File: sqlglot/parser.py
Comment (xiaohui-sun) [reply sqlglot/parser.py:6488]: Sounds great.

File: sqlglot/parser.py
Comment (xiaohui-sun) [reply sqlglot/parser.py:6518]: Updated

File: sqlglot/parser.py
Comment (xiaohui-sun) [reply sqlglot/parser.py:6537]: You are right.

File: sqlglot/generator.py
Comment (tobymao) [inline sqlglot/generator.py:3130]: why is this flat=True?

Comment (xiaohui-sun) [reply sqlglot/generator.py:3130]: So there is no need to flat since there is only one action?

Comment (tobymao) [reply sqlglot/generator.py:3130]: if it's flat, it may not be pretty printed

Comment (xiaohui-sun) [reply sqlglot/generator.py]: Removed. Although I am not very clear how this controls the output. Thanks.

Comment (xiaohui-sun) [reply sqlglot/generator.py]: got it. yes it will only generate one line with flat=True.

File: tests/test_transpile.py
Comment (georgesittas) [inline tests/test_transpile.py:848]: We shouldn't get rid of this test. Let's make sure this is still supported (either parses into an alter expression or into a command), otherwise this PR will introduce a regression.

File: tests/test_parser.py
Comment (georgesittas) [inline tests/test_parser.py:592]: I'd suggest moving these two tests to `identity.sql`, not sure why `test_rename_table` was in `test_parser.py` to begin with.

File: sqlglot/parser.py
Comment (georgesittas) [inline sqlglot/parser.py:6501]: I think this needs to be changed to `alter_token` or something, there's no `CREATE` in this context.

File: sqlglot/parser.py
Comment (georgesittas) [inline sqlglot/parser.py:6495]: Why are we passing `self._prev` to `_parse_select` here? We can also just move this inline within the `ALTER_PARSERS` dict, like we do for `"DELETE"`.

File: sqlglot/expressions.py
Comment (georgesittas) [inline sqlglot/expressions.py:4277]: Since this is now expecting a `kind` argument, so I'd rename the expression to `Alter` because `AlterTable` is a misnomer when the `kind` is not `"TABLE"`.

@tobymao FYI this is gonna be a breaking change, e.g. every `AlterTable` in SQLMesh is affected.

File: sqlglot/generator.py
Comment (georgesittas) [inline sqlglot/generator.py:3140]:
```suggestion
        kind = self.sql(expression, "kind")
```

**Review by georgesittas — COMMENTED** [review]
Hey @xiaohui-sun, did another pass on the PR. Let's please increase the test coverage for this, e.g. at least add the examples in Spark's [docs](https://spark.apache.org/docs/3.0.0-preview2/sql-ref-syntax-ddl-alter-view.html#syntax-3).

Ideally we need to investigate which dialects support `ALTER VIEW` and also add tests for them and ensure this PR doesn't break stuff.

The `ALTER ...` statement family is an important one and this is gonna be an important breaking change, so let's ensure we don't break stuff.

cc @VaggelisD to also take a look at this & other dialects if you have cycles.

**Comment by VaggelisD** [comment]
I did a pass over other dialects and it seems that `ALTER VIEW AS` is also supported by [MySQL](https://dev.mysql.com/doc/refman/8.4/en/alter-view.html) and [T-SQL](https://learn.microsoft.com/en-us/sql/t-sql/statements/alter-view-transact-sql?view=sql-server-ver16).

Both of these dialects allow options before & after the `AS <select>` clause so if we don't want to increase the PR scope, we must at least make sure that they're parsed gracefully as Commands

File: tests/test_transpile.py
Comment (xiaohui-sun) [reply tests/test_transpile.py:848]: This test is for [unsupported syntax](https://github.com/tobymao/sqlglot/pull/3873/files/08706b9b4bab7fb35e520dcb732a70ab6100501a#diff-ddc45c8cfc37c82c0e36c52bcd13d8de90b866af0a3150bdefc157729db6d63dR867). I will move it to another test.

File: tests/test_parser.py
Comment (xiaohui-sun) [reply tests/test_parser.py:592]: Sure. Will add more tests.

Comment (xiaohui-sun) [reply tests/test_parser.py]: moved.

File: sqlglot/parser.py
Comment (xiaohui-sun) [reply sqlglot/parser.py]: fixed

File: sqlglot/parser.py
Comment (xiaohui-sun) [reply sqlglot/parser.py]: renamed

File: sqlglot/generator.py
Comment (xiaohui-sun) [reply sqlglot/generator.py]: updated

**Comment by xiaohui-sun** [comment]
> I did a pass over other dialects and it seems that `ALTER VIEW AS` is also supported by [MySQL](https://dev.mysql.com/doc/refman/8.4/en/alter-view.html) and [T-SQL](https://learn.microsoft.com/en-us/sql/t-sql/statements/alter-view-transact-sql?view=sql-server-ver16).
> 
> Both of these dialects allow options before & after the `AS <select>` clause so if they're not currently supported, we must at least make sure that they're parsed gracefully as Commands

I will add more tests there.

File: tests/test_transpile.py
Comment (xiaohui-sun) [reply tests/test_transpile.py:848]: Actually I can't find any SQL dialect supports "ALTER VIEW ... ALTER COLUMN ..."
Only T-SQL supports something [similar](https://learn.microsoft.com/en-us/sql/t-sql/statements/alter-view-transact-sql?view=sql-server-ver16#syntax) but requires two statements like below:
```
ALTER VIEW view_name
ALTER COLUMN column_name datatype [NULL | NOT NULL];

```

**Comment by teja71296** [comment]
How can we give the input of a sql query to multiple lines as its not transpiling the query.

Below input which is not working.

python sqlglot/__main__.py "CREATE VIEW customer_order_data \n
AS \n
select cd.id as customer_id, \n
CONCAT(cd.last_name, cd.first_name) AS full_name \n
god.first_order_date, \n
god.most_recent_order_date, \n
coalesce(god.number_of_orders,0) as number_of_orders \n
from customer_data cd \n
left join grouped_order_data god on cd.id = god.customer_id;" --read tsql --write snowflake


If we give the input as below its working but why?

python sqlglot/__main__.py "CREATE VIEW customer_order_data AS select cd.id as customer_id, CONCAT(cd.last_name, cd.first_name) AS full_name, god.first_order_date, gd.most_recent_order_date, coalesce(god.number_of_orders,0) as number_of_orders from customer_data cd left join grouped_order_data god on cd.id = god.customer_id;" --read tsql --write snowflake

And also need help how to translate Stored procedure, from sql to Snowflake - Need help

File: sqlglot/expressions.py
Comment (xiaohui-sun) [reply sqlglot/expressions.py:4277]: I can rename to `Alter`, and that would touch lots of places since `AlterTable` is used widely.

File: tests/dialects/test_spark.py
Comment (georgesittas) [inline tests/dialects/test_spark.py:133]: Let's make all of these `validate_identity`.

**Review by georgesittas — COMMENTED** [review]
@xiaohui-sun thanks for addressing the comments! Seems like we cover several examples from the dialects that support `ALTER VIEW`, so the PR looks good. 👍

One note: it seems like there are multiple `ALTER VIEW` statements that are not fully parsed and we fallback to `Command` nodes. I noticed this in the [CI terminal output](https://github.com/tobymao/sqlglot/actions/runs/10274153786/job/28430237500?pr=3873), where you can see multiple warnings logged. Can you track these cases down and set `check_command_warning=True` in the `validate_identity` calls? We shouldn't have these warnings in the tests.

> I can rename to Alter, and that would touch lots of places since AlterTable is used widely. ([context](https://github.com/tobymao/sqlglot/pull/3873/files#r1705318614))

I think I would probably commit to this since we're adding this now anyway, so that it's implemented cleanly. I'll defer to @tobymao on this decision. TL;DR, I see 2 options:

- Rename `AlterTable` to `Alter` so that the name doesn't assume we're altering the `Table`. Since we're making a breaking change here with `kind`, this seems ok to me.
- Keep `AlterTable` and make `kind` default to `TABLE` so that this is backwards-compatible. Two downsides to this: 1) `ALTER VIEW` will be parsed into `AlterTable`, which is not a good name, 2) it's inconsistent with `exp.Create` which doesn't assume the `CREATABLE`'s kind.

**Comment by xiaohui-sun** [comment]
> we can rename

Ack. Let me rename.

File: tests/dialects/test_spark.py
Comment (xiaohui-sun) [reply tests/dialects/test_spark.py]: fixed

**Comment by xiaohui-sun** [comment]
> @xiaohui-sun thanks for addressing the comments! Seems like we cover several examples from the dialects that support `ALTER VIEW`, so the PR looks good. 👍
> 
> One note: it seems like there are multiple `ALTER VIEW` statements that are not fully parsed and we fallback to `Command` nodes. I noticed this in the [CI terminal output](https://github.com/tobymao/sqlglot/actions/runs/10274153786/job/28430237500?pr=3873), where you can see multiple warnings logged. Can you track these cases down and set `check_command_warning=True` in the `validate_identity` calls? We shouldn't have these warnings in the tests.
> 
> > I can rename to Alter, and that would touch lots of places since AlterTable is used widely. ([context](https://github.com/tobymao/sqlglot/pull/3873/files#r1705318614))
> 
> I think I would probably commit to this now so that it's implemented cleanly in one pass, but I'll defer to @tobymao on this decision. TL;DR, I see 2 options:
> 
> 1. Rename `AlterTable` to `Alter` so that the name doesn't assume we're altering the `Table`. Since we're making a breaking change here with `kind`, this seems ok to me.
> 2. Keep `AlterTable` and make `kind` default to `TABLE` so that this is backwards-compatible. Two downsides to this: 1) `ALTER VIEW` will be parsed into `AlterTable`, which is not a good name, 2) it's inconsistent with `exp.Create` which doesn't assume the `CREATABLE`'s kind.

@georgesittas Thanks for your suggestion. I have changed `AlterTable` to `Alter`, since it is a breaking change anyway.

File: sqlglot/expressions.py
Comment (xiaohui-sun) [reply sqlglot/expressions.py:4277]: Renamed.

File: sqlglot/parser.py
Comment (georgesittas) [inline sqlglot/parser.py:1019]:
```suggestion
        "AS": lambda self: self._parse_ddl_select(),
```

**Review by georgesittas — APPROVED** [review]
Thanks for the quick turnaround @xiaohui-sun - appreciate the patience!

This looks good to me, the only outstanding item is silencing the warnings. If it's a low-lift, feel free to take a stab at actually expanding the syntax support, but if it's not trivial I wouldn't worry about it.

**Comment by xiaohui-sun** [comment]
> Thanks for the quick turnaround @xiaohui-sun - appreciate the patience!
> 
> This looks good to me, the only outstanding item is silencing the warnings. If it's a low-lift, feel free to take a stab at actually expanding the syntax support, but if it's not trivial I wouldn't worry about it.

No problem. I will take a look at the warnings and add as much support as I can.

File: sqlglot/parser.py
Comment (xiaohui-sun) [reply sqlglot/parser.py]: updated

**Comment by xiaohui-sun** [comment]
I made an attempt to fix the test cases for `ALTER ... SET TABLE PROPERTIES`.
The change is more complex than I initially expected, but I'd like to get some feedback before I dive too far into it.
@georgesittas 
1. Changed `self._parse_wrapped_csv(self._parse_assignment))` to `self._parse_wrapped(self._parse_properties, optional=True)` ([link](https://github.com/tobymao/sqlglot/pull/3873/files#diff-63eb8dd82d3561bc414e3e13cf59516bd6584c5766a10072157a8a5aee4ad85fR6515)) 
2. Set the property prefix for multiple dialects.
3. Also reduced some test cases that cause warning

File: sqlglot/dialects/trino.py
Comment (georgesittas) [inline sqlglot/dialects/trino.py:40]: This doesn't look right, I think it breaks valid DDL statements that contain the `WITH (...)` syntax. See https://trino.io/docs/current/sql/create-table.html.

**Review by georgesittas — COMMENTED** [review]
Hey @xiaohui-sun, I took a quick look at the latest changes (`SET` etc) and I feel like we should avoid doing it in this PR after all, in order to avoid scope creep.

Some of the changes seem a bit sketchy, e.g. I commented on one related to the "WITH" properties in Trino. The `Property` generation override is also weird. I'm worried this isn't the right way to implement the remaining syntax and these changes may break existing DDLs.

Can you revert to the previous commit that I'd approved? It's fine if some of the `ALTER`s are commands for now, just make sure to silence the warnings using the flag I mentioned in the `validate_identity` tests. Should be good to merge after that.

File: sqlglot/dialects/trino.py
Comment (xiaohui-sun) [reply sqlglot/dialects/trino.py:40]: ok let me revert.

Comment (xiaohui-sun) [reply sqlglot/dialects/trino.py]: @georgesittas Reverted and added `check_command_warning=True`. Please take a look. Thanks

Comment (georgesittas) [reply sqlglot/dialects/trino.py]: Thanks!

**Comment by georgesittas** [comment]
Thanks for the contribution!

**Pull request merged**
